src/iceberg/manifest/manifest_reader.cc

```diff
@@ -19,61 +19,872 @@
+        ICEBERG_RETURN_UNEXPECTED(ParseDataFile(data_file_schema, view_of_column,
+                                                first_row_id, manifest_entries));
+        break;
+      }
+      default:
+        return InvalidManifest("Unsupported field: {} in manifest entry.", field_name);
+    }
+  }
+  return manifest_entries;
+}
+
+const std::vector<std::string> kStatsColumns = {"value_counts",     "null_value_counts",
+                                                "nan_value_counts", "lower_bounds",
+                                                "upper_bounds",     "record_count"};
+
+bool RequireStatsProjection(const std::shared_ptr<Expression>& row_filter,
+                            const std::vector<std::string>& columns) {
+  if (!row_filter || row_filter->op() == Expression::Operation::kTrue) {
+    return false;
+  }
+  if (columns.empty()) {
+    return false;
+  }
+  const std::unordered_set<std::string> selected(columns.cbegin(), columns.cend());
+  if (selected.contains(ManifestReader::kAllColumns)) {
+    return false;
+  }
+  if (std::ranges::all_of(kStatsColumns, [&selected](const std::string& col) {
+        return selected.contains(col);
+      })) {
+    return false;
+  }
+  return true;
+}
+
+Result<std::shared_ptr<Schema>> ProjectSchema(std::shared_ptr<Schema> schema,
+                                              const std::vector<std::string>& columns,
+                                              bool case_sensitive) {
+  if (!columns.empty()) {
+    return schema->Select(columns, case_sensitive);
+  }
+  return schema;
+}
+
+}  // namespace
+
+std::vector<std::string> ManifestReader::WithStatsColumns(
+    const std::vector<std::string>& columns) {
+  if (std::ranges::contains(columns, ManifestReader::kAllColumns)) {
+    return columns;
+  } else {
+    std::vector<std::string> updated_columns{columns};
+    updated_columns.insert(updated_columns.end(), kStatsColumns.begin(),
+                           kStatsColumns.end());
+    return updated_columns;
+  }
+}
+
+// ManifestReaderImpl constructor
+ManifestReaderImpl::ManifestReaderImpl(
+    std::string manifest_path, std::optional<int64_t> manifest_length,
+    std::shared_ptr<FileIO> file_io, std::shared_ptr<Schema> schema,
+    std::shared_ptr<PartitionSpec> spec,
+    std::unique_ptr<InheritableMetadata> inheritable_metadata,
+    std::optional<int64_t> first_row_id)
+    : manifest_path_(std::move(manifest_path)),
+      manifest_length_(manifest_length),
+      file_io_(std::move(file_io)),
+      schema_(std::move(schema)),
+      spec_(std::move(spec)),
+      inheritable_metadata_(std::move(inheritable_metadata)),
+      first_row_id_(first_row_id) {}
+
+ManifestReader& ManifestReaderImpl::Select(const std::vector<std::string>& columns) {
+  columns_ = columns;
+  return *this;
+}
+
+ManifestReader& ManifestReaderImpl::FilterPartitions(std::shared_ptr<Expression> expr) {
+  part_filter_ = expr ? std::move(expr) : True::Instance();
```

Comment:
The java implementation have 'AND': https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestReader.java